### ISSUE TITLE
revision tag warning for older versions

### ISSUE THREAD

**Issue by linsun** [istio/istio.io#9691]
Hey @Monkeyanator I wanted to get your take on this.

When following https://preliminary.istio.io/latest/docs/setup/upgrade/canary/#stable-revision-labels-experimental, and run this cmd on my 1-8 istio cluster:

```
$ istioctl x revision tag set canary --revision 1-8-0
```

I'm getting a warning that revision tag is supposed to for Istio 1.10 or newer.  Is the warning correct?  Any reason why this may not work on older version of Istio like 1.8 or 1.9 or even 1.7?

**Comment by Monkeyanator** [comment]
Yeah the reason for this is webhook patching behavior. Older istiod don't know how to patch in CA Bundles for newer mutating webhooks (different names, multiple webhooks, etc), the patching logic in 1.10 is more versatile (1.9+ actually?).

To get around this we could hack it to pull the CA bundle out of existing webhooks. There's also work to move patching to install time which would make this problem disappear

**Comment by Monkeyanator** [comment]
Agreed this command will be way more useful if we can use against arbitrarily old versions... I'll prioritize work on this (at least a flag to give the option)

**Comment by linsun** [comment]
@Monkeyanator could I ask why older Istiod need to patch CA bundles for newer mutating web hooks?

My understanding is that mutating web hooks has been part of istio-discovery for a long time so each istiod has its own mutating web hooks.  for example, istiod 1.8 has its own mutating web hook and istiod 1.10 also has its own.

**Comment by linsun** [comment]
Also, would not two istiod deployments (for example istiod 1.8 and 1.10) have the same CA bundle because they are using the same root CA among 2 revisions (base + canary)?

**Comment by Monkeyanator** [comment]
Right so to avoid having multiple istiod target the same tag webhook for patching (mutating webhook gets created when running `istioctl x tag set ...`) we made it so that istiod only patches webhooks with `istio.io/rev=<this istiod revision>`. If the revision is older, it won't know how to patch the new webhook format.

Honestly I don't have much background on whether two `istiod` revisions have to have the same CA root. If that's the case that'd probably make this simpler

**Comment by linsun** [comment]
If revision is older, if we didn't change anything from the new install, the old istiod should be able to patch the old mutating webhook config.

I think they would have same root CA bundle as both istiod (base + canary) would be using either cacerts (plugin your own key/certs) or self generated istio-ca-secret.

For example, in my env (I just installed istio 1.10 canary) and I use self signed cert:

```
$ k get mutatingwebhookconfiguration -n istio-system                  (1d2h9m) 12:48:32
NAME                                   WEBHOOKS   AGE
cert-manager-webhook                   1          15d
istio-sidecar-injector-1-10-0          2          2m15s
istio-sidecar-injector-1-8-3           1          16d

$ kubectl get secrets -n istio-system                                 (1d2h8m) 12:47:18
NAME                                               TYPE                                  DATA   AGE
default-token-wgqqr                                kubernetes.io/service-account-token   3      16d
istio-ca-secret                                    istio.io/ca-root                      5      16d
```

I've also confirmed the CABundle is the same for **both** of the istio-sidecar-injector mutatingwebhookconfigurations.
